earth_enterprise/src/common/SharedString.h

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2019 The Open GEE Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef GEO_EARTH_ENTERPRISE_SRC_COMMON_SHAREDSTRING_H_
+#define GEO_EARTH_ENTERPRISE_SRC_COMMON_SHAREDSTRING_H_
+
+#include <string>
+#include <unordered_map>
+#include <vector>
+#include <iostream>
+#include <stdint.h>
+#include <assert.h>
+
+// The goal of this class is to keep in memory only one copy of the name of
+// a cached asset or asset version. Previously, each asset stored the full names
+// of all of its children, parents, listeners, and inputs, resulting in tens of
+// thousands of duplicates of each string in very large projects.
+//
+// With this class, only an integer identifier is stored in each instance (thus
+// drastically reducing the amount of memory needed in large projects), and
+// that identifier is used to retrieve the string value when it is requested.
+// Since the string-to-identifier ratio is 1:1, equality comparisons are also 
+// much faster as we only need to compare the identifiers.
+class SharedString {
+protected:
+   class RefStorage {

```

Comment:
This is nit-picky, but SharedString is generic enough that it may not always hold asset refs. It might be a good idea to rename this (and other parts of the code that talk about refs) to StringStorage or something like that.

Reply:
Renamed as suggested.